client/state/comments/utils.js

```diff
@@ -0,0 +1,127 @@
+/**
+ * External dependencies
+ */
+import Immutable from 'immutable';
+
+/***
+ * Internal dependencies
+ */
+import {
+	COMBINED_TARGET_TYPE
+} from './constants';
+import {
+	SERIALIZE,
+	DESERIALIZE
+} from '../action-types'
+
+/***
+ * Creates a comment target id, a concatenation of siteId and postId basically
+ * @param {Number} siteId site identification
+ * @param {Number} postId post identification
+ * @returns {String} comment target id
+ */
+export function createCommentTargetId( siteId, postId ) {
+	return `${ siteId }-${ postId }`;
+}
+
+/***
+ * Creates a request id, a concatenation of siteId, postId, and query params basically
+ * @param {Number} siteId site identification
+ * @param {Number} postId post identification
+ * @param {Object} query post identification
+ * @returns {String} request id
+ */
+export function createRequestId( siteId, postId, query ) {
+	const queryKeys = Object.keys( query ).sort();
+	const queryString = queryKeys.map( ( key ) => `${ key }=${ query[key] }` ).join( '-' );
+
+	return `${siteId}-${postId}-${ queryString }`;
+}
+
+/***
+ * Creates a url string from a url template using parameters
+ * @param {String} urlTemplate a template with $var(s) to replace
+ * @param {Object} urlParams parameters for replacement in the template
+ * @param {Object} queryParams query string parameters
+ * @returns {String} url string
+ */
+export function createUrlFromTemplate( urlTemplate, urlParams, queryParams ) {
+	let constructedUrl = urlTemplate;
+
+	Object.keys( urlParams ).forEach( ( key ) => {
+		const regex = new RegExp( '\\$' + key, 'g' );
+		constructedUrl = constructedUrl.replace( regex, urlParams[ key ] );
+	} );
+
+	if ( queryParams ) {
+		constructedUrl += '?' + Object.keys( queryParams )
+				.map( ( param ) => param + '=' + encodeURIComponent( queryParams[ param ] ) )
+				.join( '&' );
+	}
+
+	return constructedUrl;
+}
+
+/***
+ * Similar to redux's combineReducers, combines a reducers to a function that reduces Immutable.Map
+ * @param {Object} reducers key - reducer name, value - reducer function
+ * @param {Function} createId id factory based on action. It should return { type: COMBINED_TARGET_TYPE, id } When the factory returns null, reducer skips processing
+ * @returns {Function} combined reducer function
+ */
+export function combineReducersWithId( reducers, createId ) {
```

Comment:
Unless we have very unique concerns, I was suggesting that we use the second data structure, and stick with Redux's [combineReducers](http://redux.js.org/docs/api/combineReducers.html) function. I'm mostly hesitant because it adds a bit of developer context switching when interacting with this subtree. Let's wait for a few more folks to chime in, before you refactor anything though. :smile: 

```
comments = { 
    tree: {
        [ `${ site_ID }_${ post_ID }` ]: { ...some comments tree... }
    },
    totalCommentsCount: {
       [ `${ site_ID }_${ post_ID }` ]: 1024
    },
    ...more props...
}
```

Also, you may find the follow section useful: https://github.com/Automattic/wp-calypso/blob/master/docs/our-approach-to-data.md#data-normalization